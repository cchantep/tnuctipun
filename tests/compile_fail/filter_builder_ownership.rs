// This test verifies that storing an intermediate filter builder in a variable
// and trying to use it later will cause ownership/borrowing compile errors

use tnuctipun::{FieldWitnesses, MongoComparable};
use tnuctipun::filters::empty;
use serde::{Serialize, Deserialize};

#[derive(Debug, Clone, Serialize, Deserialize, FieldWitnesses, MongoComparable)]
struct User {
    pub name: String,
    pub age: i32,
}

fn main() {
    // This pattern will cause ownership issues - storing intermediate builder
    let filter_builder = empty::<User>()
        .eq::<user_fields::Name, _>("John".to_string())
        .gt::<user_fields::Age, _>(18);

    // This should fail to compile due to ownership issues
    let _filter_doc = filter_builder.and(); // Error: value used after move
}
