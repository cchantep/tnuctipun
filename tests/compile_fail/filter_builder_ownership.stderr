error[E0716]: temporary value dropped while borrowed
  --> tests/compile_fail/filter_builder_ownership.rs:16:26
   |
16 |     let filter_builder = empty::<User>()
   |                          ^^^^^^^^^^^^^^^ creates a temporary value which is freed while still in use
17 |         .eq::<user_fields::Name, _>("John".to_string())
18 |         .gt::<user_fields::Age, _>(18);
   |                                       - temporary value is freed at the end of this statement
...
21 |     let _filter_doc = filter_builder.and(); // Error: value used after move
   |                       -------------- borrow later used here
   |
help: consider using a `let` binding to create a longer lived value
   |
16 ~     let mut binding = empty::<User>();
17 ~     let filter_builder = binding
   |
